provider/resource_kpi_base_search.go

```diff
@@ -1,474 +1,458 @@
 			},
 		},
 	}
 }
 
-func metric(source map[string]interface{}) interface{} {
-	m := map[string]interface{}{}
-	m["_key"] = source["id"].(string)
-	m["aggregate_statop"] = source["aggregate_statop"]
-	m["entity_statop"] = source["entity_statop"]
-	m["fill_gaps"] = source["fill_gaps"]
-	m["gap_custom_alert_value"] = source["gap_custom_alert_value"]
-	m["gap_severity"] = source["gap_severity"]
-	m["gap_severity_color"] = source["gap_severity_color"]
-	m["gap_severity_color_light"] = source["gap_severity_color_light"]
-	m["gap_severity_value"] = source["gap_severity_value"]
-	m["threshold_field"] = source["threshold_field"]
-	m["title"] = source["title"]
-	m["unit"] = source["unit"]
-	return m
-}
+func (r *resourceKpiBaseSearch) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan KpiBaseSearchState
 
-func kpiBaseSearch(ctx context.Context, d *schema.ResourceData, clientConfig models.ClientConfig) (config *models.Base, err error) {
-	body := map[string]interface{}{}
-	body["objectType"] = "kpi_base_search"
-	body["title"] = d.Get("title").(string)
-	body["description"] = d.Get("description").(string)
-	body["actions"] = d.Get("actions").(string)
-	body["alert_lag"] = d.Get("alert_lag").(string)
-	body["alert_period"] = d.Get("alert_period").(string)
-	body["base_search"] = d.Get("base_search").(string)
-	body["entity_alias_filtering_fields"] = d.Get("entity_alias_filtering_fields").(string)
-	body["entity_breakdown_id_fields"] = d.Get("entity_breakdown_id_fields").(string)
-	body["entity_id_fields"] = d.Get("entity_id_fields").(string)
-	body["is_entity_breakdown"] = d.Get("is_entity_breakdown").(bool)
-	body["is_service_entity_filter"] = d.Get("is_service_entity_filter").(bool)
-	body["metric_qualifier"] = d.Get("metric_qualifier").(string)
-
-	if d.HasChange("metrics") {
-		metricsOld, metricsNew := d.GetChange("metrics")
-		metricsOldbyId := map[string]map[string]interface{}{}
-
-		for _, metrics := range metricsOld.(*schema.Set).List() {
-			metricsData := metrics.(map[string]interface{})
-			id := metricsData["id"].(string)
-			metricsOldbyId[id] = metricsData
-		}
-		for _, metrics := range metricsNew.(*schema.Set).List() {
-			metricsData := metrics.(map[string]interface{})
-			id, ok := metricsData["id"]
-			if ok && id != "" {
-				delete(metricsOldbyId, id.(string))
-			}
-		}
-		metricsOldbyTitle := map[string]map[string]interface{}{}
-		for _, metricsData := range metricsOldbyId {
-			metricsOldbyTitle[metricsData["title"].(string)] = metricsData
-		}
-		for _, metrics := range metricsNew.(*schema.Set).List() {
-			metricsData := metrics.(map[string]interface{})
-			id, ok := metricsData["id"]
-			if !ok || id == "" {
-				title := metricsData["title"].(string)
-				if oldMetric, ok := metricsOldbyTitle[title]; ok {
-					metricsData["id"] = oldMetric["id"]
-					delete(metricsOldbyTitle, title)
-				} else {
-					metricsData["id"], _ = uuid.GenerateUUID()
-				}
```

Comment:
> It could be reached by comparing `state` and `plan` instances from the Update request.

Yeah, I think I had a very similar problem in the "Collection Data" resource, and ended up implementing the `ModifyPlan` function for it ( https://github.com/TiVo/terraform-provider-splunk-itsi/blob/main/provider/resource_collection_data.go#L524-L542 ). I think at the time, it was the only way that I found that allowed to produce concise plans, showing only the set items that actually changed, and not diffs that show the entire set changing. 

> do we really need it?

I think to start off we need to make sure we know the implications of not doing it. If i remember correctly, we did it at least so that we don't lose KPI historical data when some minor metric attribute changes (like `unit` for instance). But there might've been more to it, it's probably worth double checking. Another question that I have, how does this new implementation behave when, say, one metric out of 5 changes? Will just one ID get recomputed or all 5? Pretty sure we'd want to at least avoid the latter.

Reply:
correct, we did not want to lose historical data for the KPI.

Reply:
>  Another question that I have, how does this new implementation behave when, say, one metric out of 5 changes? Will just one ID get recomputed or all 5? 
Only one from 5, yes

Reply:
I can't tell the difference between collection data resource and this schema now. For both cases, it is nested schema. But the plan displayed properly to me, with only corrresponding metric that changed

Reply:
>I think to start off we need to make sure we know the implications of not doing it. If i remember correctly, we did it at least so that we don't lose KPI historical data when some minor metric attribute changes (like unit for instance).

We started from the statement 'if title changes - we don't want to lose historical data', afair.
Yes, I agree about the `unit`, but in this case we have not well-defined behaviour. So data that we can use as historical depends on 'math', `aggregation(aggregation(threshold_field))`. And we recomputing ids, only if title changes. But there is no straight dependency to historical data for this fields

Reply:
> I can't tell the difference between collection data resource and this schema now.

Could have to do with the fact the `id` is optional and computed in the collection data, but not sure. There were also lots of terraform/terraform framework updates since that code had been written, so it's possible something changed as well. Anyway, it's good that it's not an issue in this case.

>  We started from the statement 'if title changes - we don't want to lose historical data'
That's not what we did though. The v1 implementation works such that historical data is lost only when the title changes. I think the idea was that if the title changes, the underlying meaning of the KPI might've changed as well, so it makes sense to invalidate the historical data.

Reply:
> So data that we can use as historical depends on 'math', `aggregation(aggregation(threshold_field))`. And we recomputing ids, only if title changes.

I think the expectation was probably that if aggregations change in a meaningful matter, the developer would change the metric title as well to reflect that, in turn invalidating the historical data.

Reply:
In other words, my understanding is that in v1 we decided that we would allow users to control when the historical data should be invalidated, and that they were required to change the metric title in order to do that.

Reply:
Anyway, like i said, i think i'm okay with invalidating the entire metric if one attribute changes for now. We can always do something else in the future if we want to.